New: `prefer-promise-reject-errors` rule (fixes #7685)

not-an-aardvark opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[x] New rule

See https://github.com/eslint/eslint/issues/7685

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

This adds the `no-reject-literal` rule as discussed in https://github.com/eslint/eslint/issues/7685, and enables it in the ESLint repo for dogfooding.

Some of the logic from this rule is shared with `no-throw-literal`, so that logic has been moved out of `no-throw-literal` and into `ast-utils.js`.

**Is there anything you'd like reviewers to focus on?**

I'm wondering whether we should have the rule report object expressions, e.g.

```js
Promise.reject({ foo: 1 });
```

Advantages of reporting that case:

* `{ foo: 1 }` cannot be an instance of `Error`.

Disadvantages of reporting that case:

* `{ foo: 1 }` is not actually a *literal*, so the name of the rule might end up a bit misleading.
* Reporting that case would be inconsistent with `no-throw-literal`, which does not report `throw {foo: 1}`.

The current implementation does not report object expressions.

mysticatea commented on lib/rules/no-reject-literal.js:
```diff
@@ -0,0 +1,125 @@
+            if (resolveRejectFuncs.has(node)) {
+
+                // Get all variables in the function's scope.
+                context.getScope().variables
```
I think we can use `context.getDeclaredVariables(node)` instead of `context.getScope().variables` to simplify the finding logic (I guess `WeakSet`s are unnecessary).
If the `node` is a function, `context.getDeclaredVariables(node)` returns an array which includes the variables of the function name and parameters.

platinumazure commented on docs/rules/no-reject-literal.md:
```diff
@@ -0,0 +1,57 @@
+
+Due to the limits of static analysis, this rule cannot guarantee that you will only reject Promises with `Error` objects. While the rule will report cases where it can guarantee that the rejection reason is clearly not an `Error`, it will not report cases where there is uncertainty about whether a given reason is an `Error`. For more information on this caveat, see the [similar limitations](http: ...
+
+To avoid conflicts between rules, this rule does not report non-error values used in `throw` statements in async functions, even though these lead to Promise rejections. To lint for these cases, use the [`no-throw-literal`](http://eslint.org/docs/rules/no-throw-literal) rule instead.
```
Minor not: Consider removing "instead" at the end of this line, which may falsely imply that these rules have no point in being enabled together.

platinumazure replied:
s/not/nit, sorry.

platinumazure commented on lib/rules/no-reject-literal.js:
```diff
@@ -0,0 +1,125 @@
+            if (
+                !callExpression.arguments.length ||
+                !astUtils.couldBeError(callExpression.arguments[0]) ||
+                callExpression.arguments[0].type === "Identifier" && callExpression.arguments[0].name === "undefined"
```
I'm a bit concerned about this logic- undefined can be redefined, and further other variables could be declared-but-not-defined for the same issue. Maybe it would be better to use escope to try to find a definition with no initialization instead?

mysticatea replied:
As far as I know, other rules also do not check shadowing of the `undefined` variable to cover the old well-known pattern:

```js
;(function(undefined) {
    // do something
}())
```

platinumazure replied:
@mysticatea I have no intention of making that pattern unusable. I just want to use escope to check that whatever variable is used is definitely undefined (in value), rather than looking for the name "undefined".

platinumazure replied:
I don't see this as a blocker- let's not change this. I'll follow up with a new issue later.

platinumazure commented on lib/rules/no-reject-literal.js:
```diff
@@ -0,0 +1,125 @@
+                if (
+
+                    // Check `Promise.reject(value)` calls
+                    node.callee.type === "MemberExpression" &&
```
Can all this be extracted to one or more helper functions?

platinumazure commented on tests/lib/ast-utils.js:
```diff
@@ -904,4 +904,38 @@ describe("ast-utils", () => {
+            "foo && 2": true,
+            "foo ? 1 : 2": false,
+            "foo ? bar : 2": true,
+            "foo ? 1 : bar": true
```
Do we want to add object expressions here?

not-an-aardvark replied:
I assume you're referencing the note at the top of this PR. What do you think the function should return for object expressions?

platinumazure replied:
False, generally (my assumption is the Error constructor crafts the stack trace, so an object with a `message` or what have you will usually not suffice).

not-an-aardvark replied:
I would be fine with that behavior, but I'm wondering if it would cause confusion because the rule is called `no-reject-literal` and object expressions are not literals.

Also, I think that would be a breaking change for `no-throw-literal`, (unless we explicitly made `no-throw-literal` more lenient by allowing `ObjectExpressions`).

platinumazure commented on docs/rules/no-reject-literal.md:
```diff
@@ -0,0 +1,57 @@
+# disallow using literals as Promise rejection reasons (no-reject-literal)
+
+It is considered good practice to only instances of the built-in `Error` object for user-defined errors in Promises. `Error` objects automatically store a stack trace, which can be used to debug an error by determining where it came from. If a Promise is rejected with a non-`Error` value, it can be difficult to determine where the rejection occurred
```
I don't see a period at the end of this line, but it could be my display. Disregard if this is inaccurate.

platinumazure commented:
I can respect the possible breaking change argument. No need to change this.

On Dec 4, 2016 3:38 PM, "Teddy Katz" <notifications@github.com> wrote:

> *@not-an-aardvark* commented on this pull request.
> ------------------------------
>
> In tests/lib/ast-utils.js <https://github.com/eslint/eslint/pull/7689>:
>
> > +            foo: true,
> +            "new Foo": true,
> +            "Foo()": true,
> +            "foo`bar`": true,
> +            "foo.bar": true,
> +            "(foo = bar)": true,
> +            "(foo = 1)": false,
> +            "(1, 2, 3)": false,
> +            "(foo, 2, 3)": false,
> +            "(1, 2, foo)": true,
> +            "1 && 2": false,
> +            "1 && foo": true,
> +            "foo && 2": true,
> +            "foo ? 1 : 2": false,
> +            "foo ? bar : 2": true,
> +            "foo ? 1 : bar": true
>
> I would be fine with that behavior, but I'm wondering if it would cause
> confusion because the rule is called no-reject-literal and object
> expressions are not literals.
>
> Also, I think that would be a breaking change for no-throw-literal,
> (unless we explicitly made no-throw-literal more lenient by allowing
> ObjectExpressions).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7689>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARWen-1ezY5fjFRGGs-h-PD03ms1vU3ks5rEzLTgaJpZM4LDLYz>
> .
>

not-an-aardvark commented:
I'll add the `no-throw-literal` ObjectExpression case to our [4.0.0 to-do list](https://github.com/eslint/eslint/issues/7407).

not-an-aardvark commented:
@mysticatea I noticed you left a comment on this PR (which I think I've addressed), but you didn't leave a "Request Changes" review initially. Do you want to review this PR again, or do you think it's ready to be merged? (There's no rush, but I don't want to merge it if you still have changes to suggest.)

ljharb commented on docs/rules/no-reject-literal.md:
```diff
@@ -0,0 +1,57 @@
+
+Promise.reject(5);
+
+Promise.reject();
```
This should be totally fine - rejecting with a literal is a problem, but rejecting with nothing (simply to be able to trigger a `.catch`) is totally legit imo.

At the very least, could this be an option, so that I can use this rule while allowing `Promise.reject()`?

ljharb replied:
My use case is when I'm using promises as flow control, and I don't care about the rejection reason.

not-an-aardvark replied:
But if it's in a `.then` call, wouldn't `throw` accomplish the same thing?

To me, it seems like if you're using rejected Promises for something other than error handling, you would probably want to disable this rule anyway. But maybe I'm still misunderstanding the use case.

not-an-aardvark replied:
> There are plenty of use cases where you want to send users down the .catch path without providing an error reason

I'm still not sure I understand what these are. But even if there are valid uses for doing this, it still contradicts the point of this rule, which is to ensure that users who unexpectedly end up on the `.catch` path have useful debugging information rather than something like `Uncaught undefined`.

If someone is intentionally writing code that goes against the point of the rule, even if they have a valid reason for doing so, it seems like it would be better to just turn the rule off (possibly in a file- or line-specific manner). Adding exceptions to the rule would just cause problems for users who *do* want to ensure that Promises are only rejected with errors.

ljharb replied:
Airbnb's codebase has 102 examples of `Promise.reject()`. I somehow doubt we're the only ones using it in this fashion.

Debugging isn't the primary purpose of promises nor the most important use case.

not-an-aardvark replied:
The primary purpose of the `.reject`/`.catch` path is error handling. It is possible to use that path for control flow instead, but I don't think that's generally a good idea. For example, the synchronous equivalent is below. Its control flow unnecessarily convoluted, and could probably be refactored:

```js
try {
  // lots of code
  if (condition) {
    // jump to the error handler to start doing something else
    throw undefined;
  }
  // more code
} catch (err) {
  // do something else
}
```

I don't see why a pattern like that is any better when the control flow is asynchronous.

ljharb replied:
There are many parallels between promises and syntactic try/catch, indeed. Promises, however, are *more* powerful than that, and don't have to be limited to try/catch. What's the `Promise.all` or `Promise.race` parallel in try/catch?

not-an-aardvark replied:
I feel like we've gotten a bit off-topic. I'd be glad to keep discussing the different things that Promises can be used for and their synchronous parallels, but maybe we should do that e.g. in a chatroom so that we don't spam peoples' inboxes.

At the moment, I'm still leaning towards :-1: for adding an exception for `undefined` as Promise rejection reasons, but I'm fine with waiting to see what the rest of the team thinks about it. (Alternatively, maybe it would be better to open a separate issue after this rule lands on master, since it would be a backwards-compatible change.)

ljharb replied:
@mysticatea note that I'm not saying it should allow `Promise.reject(undefined)` - I'm saying that that is not at all the same as `Promise.reject()` because the latter has zero arguments, and the former one.

More specifically, `Promise.reject()` has no literals of any kind in it - so why is this rule, "no reject literal", concerned with it?

ljharb commented on docs/rules/no-reject-literal.md:
```diff
@@ -0,0 +1,57 @@
+Promise.reject();
+
+new Promise(function(resolve, reject) {
+  reject("something bad happened");
```
```
new Promise(function (resolve, reject) {
  throw 'something bad';
});
```
this will reject too - should it be left for `no-throw-literal`, or should this rule cover it also?

not-an-aardvark replied:
`no-throw-literal` already reports this, so I think it's probably best to not report it to avoid conflicts.

platinumazure commented:
Regarding the conversation around `Promise.reject()` with no args: There is no `throw` equivalent there, so it seems like a mistake to try to go for strict equivalency there.

To me, the rule is "When rejecting with an argument, it should be an Error so that a stack trace can be used". This is the same as no-throw-literal: "When throwing something, it should be an Error so that a stack trace can be used". But you can't throw nothing.

Even if my reasoning is not persuasive, I'd like to call out an element of ESLint philosophy, which is this: **ESLint is not opinionated**. To me, forbidding `Promise.reject()` with no args due to lack of stack trace, when that is *not* rejecting with a literal or a non-Error (except undefined, due to how that works) and when there is *no* equivalent in throw, is an *arbitrary opinion*, not (necessarily) a best practice, and I don't believe we should make that decision for the user. I think we absolutely should provide an option that lets a user opt into this behavior, since it could be seen as a *potential* best practice, but actually enforcing this in a core rule strikes me as too opinionated.

not-an-aardvark commented:
> Regarding the conversation around `Promise.reject()` with no args: There is no `throw` equivalent there, so it seems like a mistake to try to go for strict equivalency there.

This rule and `no-throw-literal` only care about the *value* of a thrown error, not the syntax that gets used to throw the error. So `throw undefined` is the `throw` equivalent to `Promise.reject()`.

> To me, forbidding `Promise.reject()` with no args due to lack of stack trace, when that is *not* rejecting with a literal or a non-Error (except undefined, due to how that works)..., is an *arbitrary opinion*.

But `Promise.reject()` *is* rejecting with a non-Error; it's equivalent to `Promise.reject(undefined)` (aside from exceptional cases where someone uses `undefined` as a variable). So it's as much of an issue as any other case reported by the rule. The goal of the rule is to prevent a Promise with anything other than an Error object. Rejecting it with `undefined` is clearly in violation of this goal, so I disagree that reporting `Promise.reject()` and `Promise.reject(undefined)` is an arbitrary opinion.

> I think we absolutely should provide an option that lets a user opt into this behavior, since it could be seen as a potential best practice, but actually enforcing this in a core rule with no way to opt out strikes me as too opinionated.

We do allow users to opt in; they can decide whether or not to enable this rule in their config. But once a user has opted into reporting non-Error rejection reasons, having the rule report `Promise.reject(undefined)` but not `Promise.reject()` based on an arbitrary syntactic distinction seems like clearly incorrect behavior.

not-an-aardvark commented:
To expand on this a bit: I acknowledge that the rule name is a bit misleading. But the documentation for `no-throw-literal` clearly indicates that the rule cares about the value of the error, not the syntax used to create that value: From the [`no-throw-literal`](http://eslint.org/docs/rules/no-throw-literal) docs:

> When it was first created, it only prevented literals from being thrown (hence the name), but it has now been expanded to only allow expressions which have a possibility of being an `Error` object.

This is why the rule currently reports statements like this:

```js
throw (condition ? "foo" : "bar");
```

Clearly, the expression `(condition ? "foo" : "bar)` is not a `Literal` node. But that's irrelevant, because it can be statically determined that the thrown value will not be an `Error` object, and so `no-throw-literal` reports it. The predicted runtime value is the important part, not the ternary expression.

Similarly, the argument (or lack thereof) in `Promise.reject()` is clearly not a `Literal` node. But that's irrelevant, because it can be statically determined that the rejection reason will not be an `Error` object, and so `no-reject-literal` reports it. Again, these rules only care about what the value will end up being at runtime; they don't care about how the exception/rejection reason is created. So having different behavior for `Promise.reject()` and `Promise.reject(undefined)` wouldn't make sense, because those are simply two different ways to create the same Promise rejection reason.

Based on the confusion in this thread, I'd like to propose two changes to this rule:

* Rename it. The other rule is clearly only called `no-throw-literal` for backwards compatibility, so I don't think we should cause further confusion by copying that name. I think something like `promise-reject-errors` would be a good name, but we can bikeshed on that.
* Disallow object and array expressions as rejection reasons, in line with @platinumazure's suggestion earlier. Again, it seems like `no-throw-literal` only has that for backwards compatibility (or maybe it's just a bug), but I think we may as well make the behavior more robust.

platinumazure commented:
I'm still not ultimately convinced that this rule should be an exact parallel to no-throw-literal. We should not be the ones who decide best practices, but rather we should defer to the community when the practices aren't objectively wrong. Thus I still think @ljharb's points deserve further consideration and should not be rejected (heh) outright.

It seems to me that if promise rejections are not the same as errors, we should not treat them the exact same in these rules (or at least, should provide options for reasonable uses).

I'm sorry to do this, but I'm now :-1: on this proposal until we can come up with something reasonable to satisfy @ljharb's use case.

not-an-aardvark commented:
> Thus I still think @ljharb's points deserve further consideration and should not be rejected (heh) outright.

I think I've given them consideration, though. The point of the proposal is to ensure that promise rejections always have stacktraces, to avoid confusing errors like [this](https://github.com/eslint/eslint/issues/7685#issuecomment-264334693). @ljharb's use case explicitly creates Promise rejections that don't have stack traces, which goes against the purpose of the rule. I don't see a need to satisfy that use case in this rule, because it gets rid of the entire benefit that would be gained by enabling the rule. Someone who wants to use Promises like that can just as easily disable this rule.

If we want to have a discussion about why using Error values as Promise rejection reasons is a good practice, then I'm certainly open to doing that. But I don't think we should make this rule worse by adding an exception for code that completely goes against what the rule is trying to enforce.

ljharb commented:
I want to be able, in the same codebase without monkeying with override comments, have `Promise.reject()`, but statically ensure that any value that *is* inside those parens is an `Error` object. I'd even want it to be stricter and not allow an *identifier* unless it was statically determined to have a local-scope-created `new Error` (or TypeError etc).

mysticatea commented:
So, is this code invalid?
I'm thinking :-1: if yes.

```js
function foo(path) {
    return new Promise((resolve, reject) => {
        fs.readFile(path, (err, content) => {
            if (err) {
                reject(err)
                return
            }
            doSomething()
        })
    })
}
```

```js
function foo(name) {
    return new Promise((resolve, reject) => {
        const cp = child_process.spawn(name)
        cp.on("error", (err) => {
            doSomething()
            reject(err)
        })
    })
}
```

ljharb commented:
Good point - rejecting with the error from a nodeback is a primary use case of `new Promise`, so perhaps my stricter behavior would only apply to `Promise.reject`.

mysticatea commented:
It sounds fair enough to me.
Hmm.

not-an-aardvark commented:
> I'd even want it to be stricter and not allow an identifier unless it was statically determined to have a local-scope-created new Error (or TypeError etc)

That would prevent things like this:

```js
function createError() {
  if (someCondition) {
    return new FooError(args);
  } else {
    return new BarError(args);
  }
}

function businessLogic() {
  if (condition) {
    return doSomethingAsync();
  } else {
    return Promise.reject(createError);
  }
}
```

But I would be fine with an option for strict reporting (where the rule reports something if it's possibly wrong, rather than certainly wrong). My main problem is allowing `Promise.reject()` without an argument, because I think that causes the same issues as `Promise.reject(notAnError);`.

mysticatea commented:
My position is:

- This rule should disallow `Promise.reject()` without an argument.
- I strongly oppose the relaxing option of `Promise.reject()`. I felt the use of `Promise.reject()` without an argument is one of bad practice since it might ignore true errors. (https://github.com/eslint/eslint/pull/7689#discussion_r91036961)  
  ESLint does not recommend a bad practice.
- I'm OK with the stricter option of `Promise.reject()`.  
  Just a concern is the pattern such as `var err = new Error(); err.code = "FOO"; Promise.reject(err)`. But we can use subclassing, so I'm not worried much.

ljharb commented:
eslint has rules that allow lots of bad practices, and leaves it up to the user whether to enable or configure the rules as such.

In the pattern `var err = new Error(); err.code = "FOO"; Promise.reject(err)` it should be trivial for eslint to know statically that `err` is in fact an error subclass created in the same scope, so that should be allowed.

If I'm explicitly enabling an option that allows `Promise.reject()`, then surely *I* know, better than the rule authors, what constitutes a "bad practice" in my codebase - that's the entire point of allowing rules to be configured.

not-an-aardvark commented:
TSC Summary: `no-reject-literal` is an accepted rule proposal that aims to prevent Promises from being rejected with non-error values. There has been considerable debate about whether the rule should have an option to allow code like `Promise.reject()` (i.e. calling `Promise.reject` without explicitly passing in an argument). On the one hand, `Promise.reject()` is arguably useful for control flow, so an option like this could allow a user to return a rejected Promise without triggering the rule. On the other hand, calling `Promise.reject()` with no arguments causes the Promise to be rejected with `undefined`, which is the same thing that the rule is trying to prevent.

TSC Question: Should we add an option to allow `Promise.reject()` without any explicit arguments?

ilyavolodin commented:
Per TSC meeting: We agreed to add a new option that would be off by default and would allow `Promise.reject()`.

not-an-aardvark commented:
As @ilyavolodin mentioned, the TSC resolution was to add an option for this, so I'll work on doing that.

Does anyone have thoughts on the two bullet points at the bottom of https://github.com/eslint/eslint/pull/7689#issuecomment-265288315? I think the rule might be less confusing if we change the name.

not-an-aardvark commented:
Since no one objected to the comment above, I'm planning to update this PR with the following changes:

* Rename the rule to `prefer-promise-reject-errors` (still up for bikeshedding)
* Make the rule disallow object literals: `Promise.reject({ foo: 1 });`
* Add an option to allow `Promise.reject()` with no arguments

not-an-aardvark commented:
Ok, I made the changes I described above. This is ready to be reviewed again.

I squashed the commits to make sure the commit message with the correct name makes it into the changelog. For reviewing convenience, [here](https://github.com/not-an-aardvark/eslint/commit/05dca086ccb621ed1c9e29c9a326d22dd068e6c2) is a copy of the changes since the last review.

gyandeeps commented on lib/rules/prefer-promise-reject-errors.js:
```diff
@@ -0,0 +1,124 @@
+                        .filter(ref => ref.isRead())
+
+                        // Only check the references that are used as the callee in a function call, e.g. `reject(foo)`.
+                        .filter(ref => ref.identifier.parent.type === "CallExpression" && ref.identifier === ref.identifier.parent.callee)
```
Just curious if this filter and above filter can be combined into single filter statement?

not-an-aardvark replied:
They could be -- do you think that would be better? Personally I think the code is clearer with the filters separated, but I'd be happy to change it if you feel differently.

platinumazure replied:
You can probably get the best of both worlds by extracting both of those filters into a single predicate function and then using that function name, if you're worried about readability and you want to follow gyandeeps' suggestion.
